### ISSUE TITLE
Bug upgrading Epinio from 1.0.0 to 1.1.0

### ISSUE THREAD

**Issue by pedro-brito-ilia** [epinio/epinio#1704]
**Describe the bug**
I'm unable to deploy any service after upgrading my Epinio instance from 1.0.0 to 1.1.0

Tried to deploy using CLI and UI, both return the same error
Tried to deploy using 1.0 and 1.1 versions of CLI, both return the same error
Tried deleting all Epinio CRDs manually, doesn't work either

**To Reproduce**

1. Install an Epinio instance on version 1.0.0
2. Upgrade to 1.1.0 (`helm upgrade epinio -n epinio epinio/epinio --version 1.1.0`)
3. Try to deploy any service (`epinio service create mysql-dev test-db`)
4. CLI returns `❌  error creating service: service create failed: Internal Server Error: error creating helm chart: create not allowed while custom resource definition is terminating`

**Logs**

On Epinio server pod logs, the return is the same: `"msg"="Error" "error"="error creating helm chart: create not allowed while custom resource definition is terminating"`

Describing the CRDs that Epinio creates for Kubernetes does not return any problems, all of them return status active an with no conflicts

**Cluster**
 - Provider: EKS
 - Kubernetes Version: 1.22
 - Load balancer: Nginx

**Additional context**
Already opened an [issue](https://github.com/epinio/epinio/issues/1702) with other problems related to the Epinio upgrade, but those were all visual and not impeditive, as with this one, I can't progress with upgrading my stack because I'm not able to deploy any service on 1.1.0

**Comment by andreas-kupries** [comment]
Your helm command (`helm upgrade epinio -n epinio epinio/epinio --version 1.1.0`) does not use `--wait`.

It is not sure that the deployment (step 2) has finished before you go and do step 3.

**Comment by andreas-kupries** [comment]
@jimmykarily believes that it probably is the [services.application.epinio.io](http://services.application.epinio.io/) crd that has the problem, as it is the service creation that fails.

**Comment by pedro-brito-ilia** [comment]
> Your helm command (`helm upgrade epinio -n epinio epinio/epinio --version 1.1.0`) does not use `--wait`.
> 
> It is not sure that the deployment (step 2) has finished before you go and do step 3.

Tried deploying again 10h+ after the upgrade, still the same error

**Comment by jimmykarily** [comment]
> > Your helm command (`helm upgrade epinio -n epinio epinio/epinio --version 1.1.0`) does not use `--wait`.
> > It is not sure that the deployment (step 2) has finished before you go and do step 3.
> 
> Tried deploying again 10h+ after the upgrade, still the same error

It's not about now needing more time to complete, it's whether it ever completed or not. You can check the status of the helm release with `helm status -n epinio epinio` (look for `STATUS`) or if you have `jq` available:

```
helm status -n epinio epinio -o json | jq .info.status
```
If it's not `deployed` then something is wrong with the upgrade.

**Comment by andreas-kupries** [comment]
# Session

Attempted to reproduce the issue, using the steps below. Failed.  This write up should hopefully make it easier to compare my actions against the original actions.

  1. Installing epinio v1.0.0, from a git checkout
  2. Basic testing of v1.0.0, by creating a service, and an application. Not bound together.
  3. Dumped the cluster state (all resources and their state)
  4. Upgraded to epinio v1.1.0, via helm command, using released charts
  5. Created another service. This went ok, contrary to expectations.
  6. Dumped the cluster state again

## Details

### v1.0.0 install

Using a git checkout, on commit e6c2b297d211ad5854f5c380128991e9ea6d2f3f (v1.0.0 tag) and using an epinio binary created from said checkout, via the following commands:

```
make acceptance-cluster-delete
make acceptance-cluster-setup
cp tmp/acceptance-kubeconfig ~/.kube/config
sleep 10
make build-images
make install-cert-manager
./scripts/prepare-environment-k3d.sh --set server.traceLevel=255
```

Post installation `epinio version` reports:

```
Kubernetes Version: v1.20.11+k3s2
Epinio Server Version: v1.0.0
Epinio Client Version: v1.0.0
```

## Creating Service and app

Using the commands

```
epinio service create mysql-dev test-db-pre-upgrade
epinio push --name fox --path assets/sample-app
```

The `assets/sample-app` is from the epinio git checkout.

Everything is ok. IOW pods and containers for app and service are created and reach state `running`.

## Cluster dump

A local script is used to determine all the resource types in the cluster, and all the resources per type.

## Upgrading

Upgrading was done by running

```
time helm upgrade epinio -n epinio epinio/epinio --version 1.1.0 --wait
```

first, then changing the checkout to commit 4809917451358f2a6d104e4ac2a5eac16324d05d (v1.1.0 tag) and rebuilding the epinio binary to match the server.

Note that I used `--wait` here. The `helm` command took about 42 seconds.

The command `epinio info`, using the remade client binary reports

```
Kubernetes Version: v1.20.11+k3s2
Epinio Server Version: v1.1.0
Epinio Client Version: v1.1.0-2022-09-05T11-32-29
```

# Created post-upgrade service

Using the command

```
epinio service create mysql-dev test-db-post-upgrade
```

No error is reported. Pods and containers for the service are created and reach state `running`.

# Second dump

The cluster resources are dumped a second time.

Comparing the CRD resources no differences are seen.

Comparing everything only expected differences are seen, i.e. new service not present before upgrade, server versions, images, etc.

**Comment by pedro-brito-ilia** [comment]
Thinking this may be a Kubernetes version compatibility issue, I re-ran my tests on a 1.20 EKS cluster and on a 1.21 cluster, again with no success.

I also ran tests on a 1.24 K3s local deployment, and it worked fine, so apparently this is directly related to the EKS deployment. Could you try reproducing this issue on AWS EKS? I am following the steps documented on https://docs.epinio.io/next/howtos/install_epinio_on_eks

**Comment by andreas-kupries** [comment]
@pgonin The CI you are referencing is not an upgrade test, only an installation test. It does not perform the thing he has trouble with. Our upgrade CI flow (https://github.com/epinio/epinio/actions/runs/2997074793) uses `k3s`. With which he has no trouble.

**Comment by enrichman** [comment]
Ok, let's then:
- extend the test to create services
- make it possible to run on different k8s versions and cloud providers
- try to reproduce the bug with the test
- fix if needed
- (optional) run schedule the test on EKS too OR extend the public cloud pipelines to also run the upgrade test

**Comment by andreas-kupries** [comment]
Reproduction achieved with GKE ...

**Comment by andreas-kupries** [comment]
The #1764 PR shows that using direct deployment of a service chart from the epinio server works across an upgrade, without an `CRD terminating` message. It seems that this is indeed an issue with the helm controller.

The referenced spike will now become the fix. (Have to add back backward compatibility, i.e. allow seeing and using the HC-based services, even if they are not created anymore).

Closed the #1749 test PR
